many: add support for /home on NFS

zyga opened the pull request:
This branch adds support for systems running on hosts with NFS-shared /home (or
fragment of /home).

The main problem is that NFS is not transparent to apparmor. When a process
attempts to access NFS-mounted files or directories it will be subject to
network mediation. In the future the kernel may understand what is going on
better and no longer require any workarounds but for now those are necessary.

With this branch, on startup only, snapd will interrogate mounted filesystems
and if NFS is found it will insert additional apparmor snippet into every
apparmor profile. This includes snap application profiles as well as the
profile of snap-confine itself.

The additional profile rules allow "network inet" and "network inet6" access.
For now we are unable to detect NFS filesystems mounted "in flight" at runtime
but we do analyze /etc/fstab and if statically declared NFS filesystem is found
there we assume it may be mounted in the future, even if it is not necessarily
mounted at the given time.

Fixes: https://bugs.launchpad.net/ubuntu/+source/snapd/+bug/1662552
Signed-off-by: Zygmunt Krynicki <zygmunt.krynicki@canonical.com>

zyga commented:
@jdstrand I've updated this based on your earlier review on IRC. This includes changes to:

- detect re-exec and skip distribution profile reload
- detect possible NFS mounts by looking at /etc/fstab (in addition to actual mounts)
- constrain the search to /home/ or subdirectories
- detect "nfs" in addition to "nfsv4"

In addition to this I added full unit test coverage and I'll re-create the spread test I lost with my snapd checkout last week.

zyga commented:
Ohhh, it's green on first push :-) Must be my lucky day

stolowski commented on interfaces/apparmor/backend.go:
```diff
@@ -63,6 +70,107 @@ func (b *Backend) Name() interfaces.SecuritySystem {
+	// Check if NFS is mounted at or under $HOME. Because NFS is not
+	// transparent to apparmor we must alter our profile to counter that and
+	// allow snap-confine to work.
+	if nfs, err := isHomeUsingNFS(); err != nil {
```
I wonder if an inability to determine if we are using NFS here should be fatal (it's fatal in this code if I'm not mistaken), unless we've full faith in our fstab/mountinfo parsing (which we probably need to have given that it's used in some other places afair). Anyway, just mentioning the concern since NFS is not a typical use case - perhaps a warning would be enough. For your consideration.

zyga replied:
Interesting question. While we don't expect any errors I agree we should stop the world about it. It is just a workaround for a missing kernel feature (NFS transparency in LSMs). I'll make this non fatal.

zyga replied:
Done now.

stolowski replied:
Thank you!

stolowski commented on interfaces/apparmor/backend.go:
```diff
@@ -63,6 +70,107 @@ func (b *Backend) Name() interfaces.SecuritySystem {
+		return false, fmt.Errorf("cannot parse /proc/self/mountinfo, %s", err)
+	}
+	for _, entry := range mountinfo {
+		if (entry.FsType == "nfs4" || entry.FsType == "nfs") && (strings.HasPrefix(entry.MountDir, "/home/") || entry.MountDir == "/home") {
```
I know we've a deeper and historical problem with hardcoded /home, but I'm not super happy about adding another hardcoded case here. Actually, if we ever fix the problem with hardcoded /home in dirs.go, this fix will break completely and this makes me a sad panda...
Would it be possible to make it more generic and future-proof (e.g. check if any of the filesystems that the snap may potentially be interested in is NFS)? Don't we get the same problem if - say - /usr or /var is NFS-mounted?

zyga replied:
This was added by a request from @jdstrand - the idea is that we want to slowly open this up to more use cases. For now we just want to support /home on NFS. We plan to expand this to CIFS as well but not in arbitrary places yet.

Note that `/home` is already explicitly described in apparmor profiles and there's no easy way around it today.

stolowski replied:
Fair enough, thanks.

stolowski commented on tests/main/nfs-support/task.yaml:
```diff
@@ -0,0 +1,96 @@
+summary: Test that snaps still work when /home is a NFS mount
```
:heart:

zyga replied:
As things go, this test obviously caught a bug in the initial code that worked fine for /home sub-directories but not the full thing :)

zyga commented on interfaces/apparmor/backend.go:
```diff
@@ -63,6 +70,107 @@ func (b *Backend) Name() interfaces.SecuritySystem {
+	policy := make(map[string]*osutil.FileState)
+
+	// Create the local policy directory if it is not there.
+	if err := os.MkdirAll(dir, 0755); err != nil {
```
Ack, will do.

zyga commented:
@chipaca oh, great idea. I'll do that.

jdstrand commented on interfaces/apparmor/backend_test.go:
```diff
@@ -492,7 +495,362 @@ func (s *backendSuite) TestSetupHostSnapConfineApparmorForReexecWritesNew(c *C)
+		{
+			fstab: "localhost:/srv/nfs /home/zyga/nfs nfs defaults 0 0",
+			nfs:   true,
+		},
```
fstab entry has 'nfs', but comment is NFSv4 here and the test below. I like these tests, we probably want tests for nfs4 and (at least spot checks for) nfs.

zyga replied:
This is how it works actually. You get NFSv4 by default but you can still specify NFSv3 by using nfsvers=3. I'll expand tests to include NFSv3 as well.

jdstrand commented on interfaces/apparmor/backend_test.go:
```diff
@@ -492,7 +495,362 @@ func (s *backendSuite) TestSetupHostSnapConfineApparmorForReexecWritesNew(c *C)
+	// We created the policy file.
+	files, err := ioutil.ReadDir(dirs.SnapConfineAppArmorDir)
+	c.Assert(err, IsNil)
+	c.Assert(files, HasLen, 1)
```
If we can't read /proc/self/exe, why would we create the policy file? We want to create the directory but I see no reason why we would fallback to creating the file, even if we don't reload, since something outside of snapd might reload the policy and then have the NFS policy included.

zyga replied:
Because the policy is equally applicable to the re-exec case. We generate the local policy file but don't reload the distro profile. When Setup is called it will choose (or not) to reload the core profile. If it chooses to reload it at that time the generated include file will be present and we will benefit from it.

In all cases I think that if we cannot examine /proc/self/exe we will have other trouble elsewhere so I think this is good as is.

jdstrand replied:
I understand with the way you wrote setupSnapConfineGeneratedPolicyImpl() why you wrote the test this way, but my point is to fail closed rather than open (and reorder the read on procSelfExe so we don't generate the policy. The idea is if there is a crazy kernel bug on reading /proc/self/exe, we fail early and loud rather than try to plod along.

I also understand your point that being able to read /proc/self/exe or not doesn't really have anything to do with isHomeUsingNFS(), which is why you wrote out the policy. I'll leave the decision up to you (ie, not a blocker).

zyga replied:
I'm changing this now.

jdstrand commented on interfaces/apparmor/backend_test.go:
```diff
@@ -492,7 +495,362 @@ func (s *backendSuite) TestSetupHostSnapConfineApparmorForReexecWritesNew(c *C)
+
+	// Setup generated policy for snap-confine.
+	err = apparmor.SetupSnapConfineGeneratedPolicy()
+	c.Assert(err, ErrorMatches, "cannot reload snap-confine apparmor profile: testing")
```
If the policy load fails, I seems to make sense to clean up the generated profile?

zyga replied:
Hmmm, I think the re-exec-from-core case may still want that but if we cannot load it ourselves then yeah, lets remove it. I'll make the adjustment.

zyga replied:
Done.

jdstrand commented on tests/main/nfs-support/task.yaml:
```diff
@@ -0,0 +1,96 @@
+execute: |
+    run_export_nfs() {
+        # NOTE: This is so silly because spellchecker complains about the real
+        # command name so w have to go and obfuscate it /o\
```
s/so w/so we/

zyga replied:
Fixed

jdstrand commented on tests/main/nfs-support/task.yaml:
```diff
@@ -0,0 +1,96 @@
+    Snapd now contains a feature where NFS-mounted /home (or any sub-directory)
+    initializes a workaround mode where all snaps gain minimal amount of network
+    permissions sufficient for NFS to operate.
+systems: [ubuntu-16.04-64]  # TODO: expand this list
```
Do you plan to address this TODO in this PR?

zyga replied:
No, it's a lot of hairy stuff already. I'll address this with separate PRs.

jdstrand commented on interfaces/apparmor/backend_test.go:
```diff
@@ -492,7 +495,362 @@ func (s *backendSuite) TestSetupHostSnapConfineApparmorForReexecWritesNew(c *C)
+	// Make it appear as if NFS was mounted under /home.
+	restore := apparmor.MockMountInfo("680 27 0:59 / /home/zyga/nfs rw,relatime shared:478 - nfs4 localhost:/srv/nfs rw,vers=4.2,rsize=524288,wsize=524288,namlen=255,hard,proto=tcp,port=0,timeo=600,retrans=2,sec=sys,clientaddr=127.0.0.1,local_lock=none,addr=127.0.0.1")
+	defer restore()
+	restore = apparmor.MockEtcFstab("")
```
I noticed that we have a lot of tests using `MockMountInfo("...")` when using `MockEtcFstab("")`, but not the other way around or when both are set (outside of `TestIsHomeUsingNFS()`). Was this intentional?

zyga replied:
I wrote mountinfo support code first and only then added the trigger code that looked for fstab. Since other tests are checking that they have equivalent power I should perhaps just say "mock stuff so that NFS workaround is enabled" without going into the details of how that is done.

zyga replied:
Also done now.

zyga commented on interfaces/apparmor/backend.go:
```diff
@@ -63,6 +70,106 @@ func (b *Backend) Name() interfaces.SecuritySystem {
+// Initialize prepares customized apparmor policy for snap-confine.
+func (b *Backend) Initialize() error {
+	// TODO: also generate policy for snap-confine from core.
+	return setupSnapConfineGeneratedPolicy()
```
I was about to say "but I really want to do that eventually" and then I realized it needs to be the way it is for now. I'll change this comment to explain why.

zyga commented on interfaces/apparmor/backend.go:
```diff
@@ -63,6 +70,106 @@ func (b *Backend) Name() interfaces.SecuritySystem {
+	// setupSnapConfineReexec below.
+	exe, err := os.Readlink(procSelfExe)
+	if err != nil {
+		return fmt.Errorf("cannot read /proc/self/exe, %s", err)
```
Done

zyga commented on interfaces/apparmor/backend.go:
```diff
@@ -63,6 +70,106 @@ func (b *Backend) Name() interfaces.SecuritySystem {
+
+	// If snapd is executing from the core snap the it means it has
+	// re-executed. In that case we are no longer using the copy of
+	// snap-confined from the host distribution but our own copy. We don't have
```
Fixed

zyga commented on interfaces/apparmor/backend.go:
```diff
@@ -63,6 +70,106 @@ func (b *Backend) Name() interfaces.SecuritySystem {
+	if output, err := exec.Command("apparmor_parser", "--replace",
+		// TODO: the name varies across distros, fix that :/
+		"--write-cache", filepath.Join(dirs.SystemApparmorDir, "usr.lib.snapd.snap-confine.real"),
+		"--cache-loc", dirs.SystemApparmorCacheDir).CombinedOutput(); err != nil {
```
Done

zyga commented on interfaces/apparmor/backend.go:
```diff
@@ -63,6 +70,106 @@ func (b *Backend) Name() interfaces.SecuritySystem {
+	}
+	fstab, err := mount.LoadProfile(etcFstab)
+	if err != nil {
+		return false, fmt.Errorf("cannot parse /etc/fstab, %s", err)
```
Done

zyga commented on interfaces/apparmor/backend.go:
```diff
@@ -254,6 +361,9 @@ func addContent(securityTag string, snapInfo *snap.Info, opts interfaces.Confine
 				// the super-broad template we are starting with.
 			} else {
 				tagSnippets = snippetForTag
+				if nfs, _ := isHomeUsingNFS(); nfs {
```
Done

zyga commented on interfaces/apparmor/template.go:
```diff
@@ -461,3 +461,13 @@ var classicJailmodeSnippet = `
+// to operate when NFS is used. This is an imperfect solution as this grants
+// some network access to all the snaps on the system.
+var nfsSnippet = `
+  # Workaround for systems using NFS, for details see:
```
Done

zyga commented:
@jdstrand I've addressed or commented on all the items you have highlighted except for the `.real` filename. I'll look into fixing that one properly now.

zyga commented on interfaces/apparmor/backend.go:
```diff
@@ -63,6 +70,106 @@ func (b *Backend) Name() interfaces.SecuritySystem {
+	// load any of the files placed in /var/lib/snapd/apparmor/snap-confine.d/.
+	// We are not using apparmor.LoadProfile() because it uses other cache.
+	if output, err := exec.Command("apparmor_parser", "--replace",
+		// TODO: the name varies across distros, fix that :/
```
Yes, I'll address this next.

zyga replied:
Done.

zyga commented:
@jdstrand I believe everything is changed or explained (in one case)

jdstrand commented:
Approved (though see my last comment (not a blocker)).

zyga commented:
@jdstrand after thinking about it I'll rework the code slightly.

niemeyer commented on interfaces/apparmor/backend.go:
```diff
@@ -63,6 +70,138 @@ func (b *Backend) Name() interfaces.SecuritySystem {
+	procSelfMountInfo = mount.ProcSelfMountInfo
+)
+
+// isHomeUsingNFS returns true if /home contains or might contain NFS mounts.
```
The "might contain" is a truism and would be nice to clarify, in the sense that /etc/fstab _might_ contain NFS entries at all times. Clearly there's something more specific implied there since it mentions both "contains" and "might contain", though.

zyga replied:
Re worded to

`// isHomeUsingNFS returns true if NFS mounts are defined or mounted under /home.`

niemeyer commented on interfaces/apparmor/backend.go:
```diff
@@ -63,6 +70,138 @@ func (b *Backend) Name() interfaces.SecuritySystem {
+}
+
+var (
+	procSelfMountInfo = mount.ProcSelfMountInfo
```
Let's please move this up to the common block above.

zyga replied:
Done

niemeyer commented on interfaces/apparmor/backend.go:
```diff
@@ -63,6 +70,138 @@ func (b *Backend) Name() interfaces.SecuritySystem {
+	}
+
+	// Check the /proc/self/exe symlink, this is needed below but we want to
+	// fail early if this fails for whatever reason.
```
Thanks for making that clear. That sort of comment helps much when someone else goes to refactor the logic.

niemeyer commented on interfaces/apparmor/backend.go:
```diff
@@ -63,6 +70,138 @@ func (b *Backend) Name() interfaces.SecuritySystem {
+func setupSnapConfineGeneratedPolicyImpl() error {
+	// Create the local policy directory if it is not there.
+	if err := os.MkdirAll(dirs.SnapConfineAppArmorDir, 0755); err != nil {
+		return fmt.Errorf("cannot create snap-confine policy directory, %s", err)
```
s/,/:/. Every error and log message in this PR seems to use this convention, and this is not our usual convention. Can you please review all error messages and make sure they follow suit?

As a more general note, let's please be careful to preserve consistency in the code base. I've seen that sort of minor discrepancy often, and if it goes unnoticed the code gradually becomes messy. We can change conventions, but we cannot change them in isolation and without notice.

zyga replied:
I was wondering why I got this wrong. I must have got the impression that the `cannot quux: %s` is the wrong way and `cannot quux, %s` is the right way of formatting errors somehow (as I tried to be consistent, just going the wrong way).

I've reviewed and corrected all the error messages around this code.

niemeyer commented on interfaces/apparmor/backend_test.go:
```diff
@@ -492,7 +495,430 @@ func (s *backendSuite) TestSetupHostSnapConfineApparmorForReexecWritesNew(c *C)
+			fstab: "localhost:/srv/nfs /home nfs4 defaults 0 0",
+			nfs:   true,
+		},
+		{
```
We tend to unify those lines elsewhere, as in `}, {`. The extra spacing doesn't buy us much. The comments can go inside the brackets instead of outside, which even improves the readability as it's closer to the data.

zyga replied:
Done, the whole section has significantly de-indented itself now.

niemeyer commented on interfaces/apparmor/export_test.go:
```diff
@@ -20,9 +20,71 @@
+	}
+	etcFstab = f.Name()
+	return func() {
+		os.Remove(etcFstab)
```
We need a check here that panics if this is /etc/fstab. We don't want to kill the system by mistake.

zyga replied:
Oh, good idea. Done

niemeyer commented on interfaces/apparmor/export_test.go:
```diff
@@ -20,9 +20,71 @@
+
+// MockEtcFstab mocks content of /etc/fstab read by isHomeUsingNFS
+func MockEtcFstab(text string) (restore func()) {
+	old := procSelfMountInfo
```
Bad copy & paste here.

zyga replied:
Ouch, thank you for noticing that. Corrected.

niemeyer commented on interfaces/apparmor/export_test.go:
```diff
@@ -20,9 +20,71 @@
+	old := procSelfMountInfo
+	f, err := ioutil.TempFile("", "fstab")
+	if err != nil {
+		panic(fmt.Errorf("cannot open temporary file %s", err))
```
s/file/file:/

zyga replied:
Done

niemeyer commented on interfaces/apparmor/export_test.go:
```diff
@@ -20,9 +20,71 @@
+		panic(fmt.Errorf("cannot open temporary file %s", err))
+	}
+	if err := ioutil.WriteFile(f.Name(), []byte(text), 0644); err != nil {
+		panic(fmt.Errorf("cannot write mock fstab file %s", err))
```
s/file/file:/

zyga replied:
Done

niemeyer commented on interfaces/apparmor/export_test.go:
```diff
@@ -20,9 +20,71 @@
+		panic(fmt.Errorf("cannot open temporary file %s", err))
+	}
+	if err := ioutil.WriteFile(f.Name(), []byte(text), 0644); err != nil {
+		panic(fmt.Errorf("cannot write mock mountinfo file %s", err))
```
s/file/file:/

zyga replied:
Done

niemeyer commented on interfaces/apparmor/export_test.go:
```diff
@@ -20,9 +20,71 @@
+	old := procSelfMountInfo
+	f, err := ioutil.TempFile("", "mountinfo")
+	if err != nil {
+		panic(fmt.Errorf("cannot open temporary file %s", err))
```
s/file/file:/

zyga replied:
Done

niemeyer commented on interfaces/apparmor/template.go:
```diff
@@ -478,3 +478,13 @@ var classicJailmodeSnippet = `
+
+// nfsSnippet contains extra permissions necessary for snaps and snap-confine
+// to operate when NFS is used. This is an imperfect solution as this grants
+// some network access to all the snaps on the system.
```
I'm honestly quite surprised that the _application_ needs network access to access the filesytem. I won't question that as we'd not have gotten that far if that wasn't true, but I'd like to hear more about how this ended up like this. It seems to make no sense to me at least.

zyga replied:
I'll find the relevant apparmor bug and link it here.

zyga replied:
I also added the reference to the bug report into the source code.

niemeyer commented on overlord/ifacestate/helpers.go:
```diff
@@ -74,13 +74,19 @@ func (m *InterfaceManager) addInterfaces(extra []interfaces.Interface) error {
 }
 
-func (m *InterfaceManager) addBackends(extra []interfaces.SecurityBackend) error {
+func (m *InterfaceManager) addAndInitBackends(extra []interfaces.SecurityBackend) error {
```
It's actually init and then add, not the opposite, but instead of fixing that let's revert and preserve the original function name here as it's symmetrical with other functions around it. It will do whatever it needs to do for adding the backends. We don't need to call out every operation in its name.

zyga replied:
Done

niemeyer commented on tests/main/nfs-support/task.yaml:
```diff
@@ -0,0 +1,136 @@
+execute: |
+    run_export_nfs() {
+        # NOTE: This is so silly because spellchecker complains about the real
+        # command name so we have to go and obfuscate it /o\
```
Ouch. If we end up with more of those, we may as well drop spellchecker.

zyga replied:
It would be good to be able to teach the spellchecker about new words. I'll see if I can do that instead of this hack.

zyga replied:
Ha, that was actually easier than I thought. Done now.

niemeyer commented on tests/main/nfs-support/task.yaml:
```diff
@@ -0,0 +1,136 @@
+    # Unmount NFS mount over /home if one exists.
+    umount /home || true
+
+    # Restore the fstab backup file if one exits.
```
exists

zyga replied:
Doen

niemeyer commented on tests/main/nfs-support/task.yaml:
```diff
@@ -0,0 +1,136 @@
+    ensure_normal_perms
+
+    # Back up the /etc/fstab file and define a NFS mount mount there.
+    cp /etc/fstab fstab.bak
```
s/bak/orig/

zyga replied:
Done

niemeyer commented on tests/main/nfs-support/task.yaml:
```diff
@@ -0,0 +1,136 @@
+
+    # Restore the fstab backup file if one exits.
+    if [ -e fstab.bak ]; then
+        mv fstab.bak /etc/fstab
```
s/bak/orig/

zyga replied:
done

niemeyer commented on tests/main/nfs-support/task.yaml:
```diff
@@ -0,0 +1,136 @@
+
+    # Install NFS server and a simple shell snap.
+    distro_install_package nfs-kernel-server
+    distro_install_package bsdgames
```
The package installations should happen at the global location next to every other package installation, and it doesn't have to be purged at restore time. Please coordinate with Sergio so that these packages also go into the image maintenance spread runs so that once we finally manage to cook packages into the images this will be there too.

zyga replied:
I'm coordinating this with Sergio now.

niemeyer commented on interfaces/apparmor/backend.go:
```diff
@@ -121,7 +260,7 @@ func setupSnapConfineReexec(snapInfo *snap.Info) error {
 	// profiles to compile but distribution package may not yet contain this
 	// directory.
-	if err := os.MkdirAll(dirs.SnapAppArmorConfineDir, 0755); err != nil {
+	if err := os.MkdirAll(dirs.SnapConfineAppArmorDir, 0755); err != nil {
```
That name makes more sense indeed, thanks.

niemeyer commented on dirs/dirs.go:
```diff
@@ -176,9 +176,9 @@ func SetRootDir(rootdir string) {
 	SnapDataDir = filepath.Join(rootdir, "/var/snap")
 	SnapDataHomeGlob = filepath.Join(rootdir, "/home/*/snap/")
 	SnapAppArmorDir = filepath.Join(rootdir, snappyDir, "apparmor", "profiles")
+	SnapConfineAppArmorDir = filepath.Join(rootdir, snappyDir, "apparmor", "snap-confine.d")
```
Can we please fix the name for this directory to "snap-confine" while we have time (no .d)? That convention doesn't make much sense here.

zyga replied:
Done

niemeyer commented on interfaces/apparmor/backend.go:
```diff
@@ -63,6 +70,138 @@ func (b *Backend) Name() interfaces.SecuritySystem {
+	return setupSnapConfineGeneratedPolicy()
+}
+
+// setupSnapConfineGeneratedPolicyImpl inspects the system and sets up local
```
That's three levels of indirection, and looking through the code it doesn't look like we need any of it. Can we please have the logic below under the real Initialize, and use the real initialize from tests?

zyga replied:
Sure, done now. This also removed some extra boiler plate code :)

zyga commented:
With the exception of unified location of package installation this is all done now. I've also renamed the generated file to `nfs-support` and changed glob to `*`. Let's see if I missed anything or if tests will work.

gdevenyi commented:
This fix does not handle autoFS mounted NFS home directories, see
https://bugs.launchpad.net/ubuntu/+source/snapd/+bug/1782873
https://bugs.launchpad.net/ubuntu/+source/snapd/+bug/1784774

gdevenyi commented:
To be clear, this fails to address AutoFS home directories because snapd starts up on boot and sees no NFS directories mounted. If I then login an NFS user, and restart snapd, it enables NFS support.

zyga commented:
@gdevenyi thank you for this information. Could you please open a bug report on https://bugs.launchpad.net/snapd/+filebug - I will look at fixing it for 2.40 if time permits.

As I'm not using autofs myself any information that would help me to reproduce the problem would be of tremendous help.

chipaca commented:
@zyga, autofs is covered by the "systems where users disappear if not logged in" caveat

zyga commented:
@chipaca can you tell me more? Perhaps I know incorrectly but my vague memory of autofs is just something that mounts on demand.

chipaca commented:
exactly, it mounts `/home/<user>/` when they log in, and unmounts it when they log out. That's why it's explicitly called out in [limitations of snapd](https://forum.snapcraft.io/t/limitations-in-snapd/9718) (well, automount, which is ... wait, automount and autofs are the same thing, right?

chipaca commented:
> automount is the program used to configure a mount point for autofs. When autofs is started, an automount daemon is spawned for each map.

gdevenyi commented:
https://bugs.launchpad.net/snapd/+bug/1821193

masip85 commented:
Isn't this issue fixed yet? According to: https://forum.snapcraft.io/t/limitations-in-snapd/9718 is not fixed.

I mount my secondary data disk with /etc/fstab:
`/dev/sda2 /mnt/sda2 auto nosuid,nodev,nofail,x-gvfs-show 0 0`
 $HOME path is: 
`/mnt/sda2/home/user/ `

Any workaround to have my $HOME path in my secondary disk and make my snap apps work?

chipaca commented:
@masip85 add an entry to /etc/fstab (or a mount unit) to bind-mount `/mnt/sda2/home` to `/home` ?

masip85 commented:
> add an entry to /etc/fstab

isn't that what happens when I change home folder with usermod?
